### ISSUE TITLE
WebMock integration is broken

### ISSUE THREAD

**Issue by ixti** [httprb/http#212]
When WebMock used (particulary VCR), first, not yet, recorded request fails with:

```
HTTP::StateError: body has already been consumed
```

**Comment by tarcieri** [comment]
I assume you fixed this @ixti, because I have been using this gem with Webmock without issue for quite some time.

**Comment by ixti** [comment]
Well I didn't. But after thinking sometime how problem could have disappeared, I eralized where it was happening. And it turns out that it was you who fixed the issue... By dropping out caching...

**Comment by ixti** [comment]
In other words. This should be considered as fixed/closed, yes. :D

**Comment by grosser** [comment]
still seeing this with http (2.0.3) + webmock (2.1.0) ... removing webmock makes the tests work ... (it is in allow_web_connect mode for this set of tests)

**Comment by grosser** [comment]
tried simply disabling that check and it did not blow up, but no longer had responses ...

```Ruby
HTTP::Response::Body.class_eval do
  undef stream!
  def stream!
    @streaming = true
  end
end
```

solved by using 

```Ruby
before { WebMock.disable! } # cannot use allow_web_connect! see https://github.com/httprb/http/issues/212
```

**Comment by tarcieri** [comment]
@grosser do you have a minimal repro? I'm not encountering this problem with the same versions in my projects that use webmock with http.rb.

**Comment by grosser** [comment]
here we go :D 

... not sure if that helps ... replacing with just a Http.get did not trigger it ... 

```Ruby
# Gemfile
source 'https://rubygems.org'

ruby File.read('.ruby-version').strip

gem 'kubeclient', git: 'https://github.com/grosser/kubeclient.git', branch: 'grosser/httpv1'
gem 'rack'
gem 'webmock'

# code
require 'bundler/setup'
require 'minitest/autorun'
require 'kubeclient'
require 'rack'
require 'open-uri'
require 'json'
require 'webmock/minitest'

# frozen_string_literal: true
class FakeServer
  def self.open(port, replies)
    server = new(port, replies)
    server.boot
    yield server
  ensure
    server.shutdown
  end

  def initialize(port, replies)
    @port = port
    @replies = replies
  end

  def boot
    Thread.new do
      Rack::Handler::WEBrick.run(
        self,
        Port: @port,
        Logger: WEBrick::Log.new("/dev/null"),
        AccessLog: []
      ) { |s| @server = s }
    end
  end

  def wait
    Timeout.timeout(10) do
      loop do
        begin
          socket = TCPSocket.new('localhost', @port)
          socket.close if socket
          sleep 0.1 if ENV['CI']
          return
        rescue Errno::ECONNREFUSED
          nil
        end
      end
    end
  end

  def call(env)
    path = env.fetch("PATH_INFO")
    code, reply = @replies[path]
    unless code
      puts "ERROR: Missing reply for path #{path}" # kubeclient does not show current url when failing
      raise
    end
    # adding \n to simulate the stream ... maybe array here ?
    reply = [reply] unless reply.is_a?(Array)
    [code, {}, reply.map { |r| r.to_json << "\n" }]
  end

  def shutdown
    @server.shutdown if @server
  end
end

describe 'foobar' do
  describe "legacy tests, do not add here" do
    before { WebMock.allow_net_connect! } # cannot use allow_web_connect! see https://github.com/httprb/http/issues/212

    it "watches container lifecycle" do
      FakeServer.open(9000, {'/api/v1/watch/events' => [200, {}]}) do |kube|
        kube.wait
        Kubeclient::Client.new('http://localhost:9000/api', "v1").watch_events(field_selector: 'involvedObject.kind=Pod').each {  }
      end
    end
  end
end
```

**Comment by chamnap** [comment]
I'm still facing this issue with webmock (3.1.1) and http (3.0.0) too. Thanks @grosser for the workaround! 👍

**Comment by wflanagan** [comment]
Problem still exists at the time of this writing.

**Comment by phoet** [comment]
with webmock 3.5.1 this is still an issue: https://github.com/penseo/firefighter/pull/3/files#diff-d2b7cb17e4aea0d9285eca5318ae8885R22

**Comment by tarcieri** [comment]
Anyone know why it broke this time?

**Comment by phoet** [comment]
@tarcieri has it been fixed in the mean time?

**Comment by tarcieri** [comment]
Webmock integration has been broken and fixed a half dozen times now. Perhaps with #499 it could be implemented in a less brittle way.

**Comment by phoet** [comment]
@tarcieri i'm asking because this particular issue is open since 2016

**Comment by phoet** [comment]
wouldnt it make sense to have a test-harness using webmock to check if it still works in order to limit the number of bugs occuring after refactorings and such?

**Comment by tarcieri** [comment]
That sounds like a good idea

**Comment by ixti** [comment]
Integration is on webmock's side, thus having CI for this on our end seems a bit weird to me. Thus I believe such regression test should be added to webmock's integration instead.

**Comment by phoet** [comment]
just guessing here, but i think that one of the problems is probably that webmocks "integration" is some kind of monkeypatch on httprb internals because there is not a proper API for those kind of use-cases.

if it seems weird to you, it could be an option to improve the test-support on the webmock side. i dont know how they handle their development and what methods they use in order to keep track with all the supported http libraries.

**Comment by ixti** [comment]
Basically they just have a different shims and specs for those. :D In any case, I guess it would be awesome to start with test case that will reproduce this bug with using HTTP gem only. That will help a lot.

**Comment by tarcieri** [comment]
@phoet I'd agree the reason they keep breaking is the way they're presently implemented is quite brittle and easily broken

> because there is not a proper API for those kind of use-cases.

I think #499 is the basis of such an API /cc @paul

**Comment by ixti** [comment]
Without monkey-patches we will have to provide a global middlewares (same way as #499). And this will be even more brittle, as for WebMock shimming must be as close to metal as possible to avoid stubbing the world.

**Comment by jrochkind** [comment]
WebMock (3.13) currently still seems to break any use of *streaming* in http-rb (any version really). I got bit by it, and it took me a while to figure out what was going on. It is a big gotcha.

**Comment by bblimke** [comment]
I have applied the fix in WebMock master branch https://github.com/bblimke/webmock/commit/216cad94dff5dc03a7db21f9d7c2c98fa8f222f1
Big kudos to @aka-nez for the [issue report](https://github.com/bblimke/webmock/issues/1017) and the sample code provided – it was super helpful in nailing this down.

The fix is a bit of a workaround. When WebMock is switched on, it'll swap out the original Response#body for a new one that can be read or streamed again. I know it's not the best fix, as it changes the original response behavior, but it was the best I could come up with given that the original Response::Body can't be reset.

**Comment by bblimke** [comment]
The fix is now released in WebMock 3.23.0

**Comment by ixti** [comment]
@bblimke thank you!
